infrastructure/src/main/java/org/corfudb/infrastructure/CorfuMsgHandler.java

```diff
@@ -209,43 +211,36 @@ private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
     private Handler<CorfuMsg> generateConditionalHandler(@Nonnull final AbstractServer server,
             @Nonnull final CorfuMsgType type,
             @Nonnull final Handler<CorfuMsg> handler) {
-        if (!MetricsUtils.isMetricsCollectionEnabled()) {
-            // If metrics are disabled, register the handler without instrumentation.
-            return (msg, ctx, r) -> {
-                if (server.isShutdown()) {
-                    log.warn("Server received {} but is shutdown.", msg.getMsgType().toString());
-                    r.sendResponse(ctx, msg, CorfuMsgType.ERROR_SHUTDOWN_EXCEPTION.msg());
-                    return;
-                }
-
-                if (!server.isServerReadyToHandleMsg(msg)) {
-                    r.sendResponse(ctx, msg, CorfuMsgType.NOT_READY.msg());
-                    return;
-                }
+        // Generate a timer based on the Corfu message type
+        final Timer timer = getTimer(type);
+
+        // Register the handler. Depending on metrics collection configuration by MetricsUtil,
+        // handler will be instrumented by the metrics context.
+        return (msg, ctx, r) -> {
+            if (server.isShutdown()) {
+                log.warn("Server received {} but is shutdown.", msg.getMsgType().toString());
+                r.sendResponse(ctx, msg, CorfuMsgType.ERROR_SHUTDOWN_EXCEPTION.msg());
+                return;
+            }
+
+            if (!server.isServerReadyToHandleMsg(msg)) {
+                r.sendResponse(ctx, msg, CorfuMsgType.NOT_READY.msg());
+                return;
+            }
 
+
+            try (Timer.Context context = MetricsUtils.getConditionalContext(timer)) {
```

Comment:
According to me, this would have a performance impact on every incoming RPC even if metrics collection is disabled. The getConditionalContext would get the current time which would be a system call.

Reply:
@zalokhan The current behavior of getConditionalContext is different that is when the metric collection is disabled no timer will be assigned and there won't be such invocation for current time from that path. https://github.com/CorfuDB/CorfuDB/blob/a6cd0a3f5a1e0ebedd3ee4e00153ef1c684ce06a/runtime/src/main/java/org/corfudb/util/MetricsUtils.java#L218
Similarly there won't be any invocation when stopping a conditional context. Let me know If there are other potential paths that lead to mentioned timing invocations.